man/pn.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PN 1"
.TH PN 1 "2014-09-13" "Clockwork v2.3.0" "Clockwork Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
pn \- Pendulum Interpreter
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpn\fR [options] script.pn
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBClockwork\fR is a configuration management system designed to
securely and correctly enforce configuration policies on lots of
hosts.
.PP
It can ensure that files have the right attributes (owner,
group, permissions, etc.) and content.  It can maintain a specific
set of installed packages, and even make sure that system user
accounts exist.
.PP
\&\fBpn\fR is a standalone interpreter for the Pendulum language,
which is what \fBcogd\fR uses internally to apply policies to the
local machine.
.PP
\&\fBpn\fR is not for everyone.  In fact, most people will never need
the run snippets of Pendulum code outside the scope of \fBcogd\fR.
\&\fBpn\fR really shines when used to aide in debugging Clockwork,
and as an exploratory utility for writing new resource types.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-V\fR, \fB\-\-version\fR" 4
.IX Item "-V, --version"
Print version and copyright information, and then exit.
.IP "\fB\-v\fR, \fB\-\-verbose\fR" 4
.IX Item "-v, --verbose"
Increase logging verbosity by one level.
.IP "\fB\-q\fR, \fB\-\-quiet\fR" 4
.IX Item "-q, --quiet"
Disable logging.
.IP "\fB\-T\fR, \fB\-\-trace\fR" 4
.IX Item "-T, --trace"
Enable \fB\s-1TRACE\s0\fR mode while executing the script.
In this mode, \fBpn\fR will print debugging messages to standard
error that should help to illustrate what opcodes it is executing,
what user-defined functions are doing, branching decisions and more.
.IP "\fB\-\-syslog\fR \fI\s-1FACILITY\s0\fR" 4
.IX Item "--syslog FACILITY"
Redirect all output to syslog, using the target facility.
.IP "\fB\-\-debug\fR" 4
.IX Item "--debug"
Enable debug-level logging.  Useful for development, but not much else.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBclockwork\fR(7), \fBclockd\fR(1), \fBclockd.conf\fR(5) and \fBcogd.conf\fR(5)
.SH "AUTHOR"
.IX Header "AUTHOR"
Clockwork was designed and written by James Hunt.


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




src/pn.c
/*
  Copyright 2011-2014 James Hunt <james@jameshunt.us>

  This file is part of Clockwork.

  Clockwork is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Clockwork is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Clockwork.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <syslog.h>
#include <getopt.h>

#include "clockwork.h"
#include "pendulum.h"
#include "pendulum_funcs.h"

int main(int argc, char **argv)
{
	int trace = 0, check = 0, debug = 0, saferun = 0;
	int log_level = LOG_WARNING;
	const char *ident = "pn", *facility = "stdout";

	const char *short_opts = "h?vqDVTc";
	struct option long_opts[] = {
		{ "help",     no_argument,       NULL, 'h' },
		{ "verbose",  no_argument,       NULL, 'v' },
		{ "quiet",    no_argument,       NULL, 'q' },
		{ "debug",    no_argument,       NULL, 'D' },
		{ "version",  no_argument,       NULL, 'V' },
		{ "trace",    no_argument,       NULL, 'T' },
		{ "check",    no_argument,       NULL, 'c' },
		{ "syslog",   required_argument, NULL, '0' },
		{ "nofork",   no_argument,       NULL, '1' },

		{ 0, 0, 0, 0 },
	};

	int opt, idx = 0;
	while ( (opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != -1) {
		switch (opt) {
		case 'h':
		case '?':
			printf("USAGE: pn [OPTIONS] script.pn\n");
			exit(0);

		case 'v':
			log_level++;
			break;

		case 'q':
			log_level = 0;
			debug = 0;
			break;

		case 'D':
			debug = 1;
			break;

		case 'V':
			printf("pn (Clockwork) %s runtime v%04x\n"
			       "Copyright (C) 2014 James Hunt\n",
			       PACKAGE_VERSION, PENDULUM_VERSION);
			exit(0);

		case 'T':
			trace = 1;
			break;

		case 'c':
			check = 1;
			break;

		case '0': /* --syslog */
			facility = argv[optind];
			break;

		case '1': /* --nofork */
			saferun = 0;
			break;
		}
	}

	if (optind != argc - 1) {
		printf("USAGE: pn [OPTIONS] script.pn\n");
		exit(0);
	}

	if (debug) log_level = LOG_DEBUG;
	cw_log_open(ident, facility);
	cw_log_level(log_level, NULL);

	FILE *io;
	if (strcmp(argv[optind], "-") == 0) {
		cw_log(LOG_INFO, "reading from standard input");
		io = tmpfile();
		if (!io) {
			cw_log(LOG_ERR, "failed to create temporary file: %s", strerror(errno));
			return 2;
		}
		char buf[8192];
		size_t nread, nwrit;
		for (;;) {
			nread = fread(buf, 1, 8192, stdin);
			cw_log(LOG_DEBUG, "read %li bytes from stdin", nread);
			if (nread <= 0) {
				if (feof(stdin)) break;
				cw_log(LOG_ERR, "read failed: %s", strerror(errno));
				fclose(io);
				return 2;
			}

			nwrit = fwrite(buf, 1, nread, io);
			cw_log(LOG_DEBUG, "wrote %li bytes to temporary file", nwrit);
			if (nwrit != nread) {
				cw_log(LOG_ERR, "failed to write to temp file: %s", strerror(errno));
				fclose(io);
				return 2;
			}
		}
	} else {
		cw_log(LOG_INFO, "reading from %s", argv[optind]);
		io = fopen(argv[optind], "r");
		if (!io) {
			fprintf(stderr, "Failed to open %s: %s\n",
					argv[optind], strerror(errno));
			return 2;
		}
	}

	pn_machine m;
	pn_init(&m);
	m.trace = trace;

	/* register clockwork pendulum functions */
	pendulum_init(&m, NULL);

	/* parse and run the input script */
	int rc = pn_parse(&m, io);
	if (rc != 0) {
		fprintf(stderr, "Syntax error: %s\n", strerror(errno));
		return 1;
	}
	if (check)
		printf("Syntax OK\n");
	else
		rc = saferun ? pn_run_safe(&m) : pn_run(&m);

	pendulum_destroy(&m);
	pn_destroy(&m);
	return rc;
}
